core/sequencer/beam.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// To exclude beam dependencies run go with --tag nobeam.
+// +build !nobeam
+
+package sequencer
+
+import (
+	"context"
+
+	"github.com/apache/beam/sdks/go/pkg/beam"
+	"github.com/apache/beam/sdks/go/pkg/beam/x/beamx"
+
+	spb "github.com/google/keytransparency/core/sequencer/sequencer_go_proto"
+	tpb "github.com/google/trillian"
+)
+
+// CreateRevisionWithBeam is an implementation of CreateRevFn using beam.
+func CreateRevisionWithBeam(ctx context.Context, directoryID string, rev int64,
+	metaProto *spb.MapMetadata, batchSize int32,
+	mw MapWriter, mr MapReader, lr LogReader) error {
+
+	p := beam.NewPipeline()
+	scope := p.Root()
+
+	meta := beam.Create(scope, metaProto)
+	dirID := beam.Create(scope, directoryID)
+	// Read each logID in parallel.
+	sourceSlices := beam.ParDo(scope, splitMeta, meta) // KV<logID, source>
+	logItems := beam.ParDo(scope, lr.ReadLog, sourceSlices,
+		beam.SideInput{Input: dirID},
+		beam.SideInput{Input: beam.Create(scope, batchSize)}) // *ktpb.EntryUpdate
+
+	keyedMutations := beam.ParDo(scope, mapLogItem, logItems) // KV<index, *ktpb.EntryUpdate>
+
+	// Read the map
+	indexes := beam.Combine(scope, &mergeIndexFn{}, beam.DropValue(scope, keyedMutations)) // []index
+	mapLeaves := beam.ParDo(scope, mr.ReadMap, indexes,
```

Comment:
And then this line does only one call to `ReadMap`, and puts the full list of indices into it?

Reply:
As discussed in person, this is indeed the case, so KT instance that processes this one []index item is a bottleneck in the pipeline. There should be some sharding in place instead -- could be implemented with `beam.Partition`, for example.

The downside of `Partition` is that the number of shards has to be a "constant" that can't depend on the input `PCollection`, so we can't decrease the number of shards as we get less input. But it can be a flag, and maybe it's okay to restart the binary when we expect the load to change.

Also, I am not sure it's impossible to "block" on `PCollection` and read the value from it in the middle of the pipeline. If it is possible then we can use `Partition` with input-dependent number of shards.

Reply:
Finally, note that we could implement our own version of `Partition` that takes `N` as a side input.